kratos/solving_strategies/convergencecriterias/displacement_criteria.h

```diff
@@ -280,40 +287,74 @@ class DisplacementCriteria : public ConvergenceCriteria< TSparseSpace, TDenseSpa
     ///@name Member Variables
     ///@{
 
-    TDataType mRatioTolerance;
+    TDataType mRatioTolerance; /// The ratio threshold for the norm of the residual
 
-    TDataType mAlwaysConvergedNorm;
+    TDataType mAlwaysConvergedNorm; /// The absolute value threshold for the norm of the residual
+
+    TDataType mReferenceDispNorm; /// The ratio threshold for the norm of the residual
 
-    TDataType mReferenceDispNorm;
 
     ///@}
     ///@name Private Operators
     ///@{
 
+    /**
+     * @brief This method computes the reference norm
+     * @details It checks if the dof is fixed
+     * @param rDofSet Reference to the container of the problem's degrees of freedom (stored by the BuilderAndSolver)
+     */
     void CalculateReferenceNorm(DofsArrayType& rDofSet)
     {
-        mReferenceDispNorm = TDataType();
-        TDataType temp;
-
-        //// x is set to its value at the beginning of the time step
-        //typename DofsArrayType::iterator it2;
-        //for (it2=rDofSet.begin();it2 != rDofSet.end(); ++it2)
-        //{
-        //	if ( !(*it2)->IsFixed()  )
-        //	{
-        //		temp = mpModel->Value(*it2);
-        //		mReferenceDispNorm += temp*temp;
-        //	}
-        //}
-        for(typename DofsArrayType::iterator i_dof = rDofSet.begin() ; i_dof != rDofSet.end() ; ++i_dof)
-        {
-            if(i_dof->IsFree())
-            {
-                temp = i_dof->GetSolutionStepValue();
-                mReferenceDispNorm += temp*temp;
+        TDataType reference_disp_norm = TDataType();
+        TDataType dof_value;
+
+        #pragma omp parallel for reduction(+:reference_disp_norm)
+        for (int i = 0; i < static_cast<int>(rDofSet.size()); i++) {
+            auto it_dof = rDofSet.begin() + i;
+
+            if(it_dof->IsFree()) {
+                dof_value = it_dof->GetSolutionStepValue();
+                reference_disp_norm += dof_value * dof_value;
+            }
+        }
+        mReferenceDispNorm = std::sqrt(reference_disp_norm);
+    }
+
+    /**
+     * @brief This method computes the final norm
+     * @details It checks if the dof is fixed
+     * @param rDofNum The number of DoFs
+     * @param rDofSet Reference to the container of the problem's degrees of freedom (stored by the BuilderAndSolver)
+     * @param Dx Vector of results (variations on nodal variables)
+     */
+    TDataType CalculateFinalCorrectionNorm(
+        SizeType& rDofNum,
+        DofsArrayType& rDofSet,
+        const TSystemVectorType& Dx
+        )
+    {
+        // Initialize
+        TDataType final_correction_norm = TDataType();
+        SizeType dof_num = 0;
+
+        // Loop over Dofs
+        #pragma omp parallel for reduction(+:final_correction_norm,dof_num)
+        for (int i = 0; i < static_cast<int>(rDofSet.size()); i++) {
+            auto it_dof = rDofSet.begin() + i;
+
+            IndexType dof_id;
+            TDataType variation_dof_value;
+
+            if (it_dof->IsFree()) {
```

Comment:
i just realized that if the dof_is_fixed than the dx is zero, so it was not harming not to do this check :-(

Reply:
In the displacement criteria to ponderate the number of dofs is considered, then if affects

Reply:
how is this different from the previous funciton?